wrote README.md

RohanBhirangi opened the pull request:
Added basic README.md
Describes in-toto and gives a simple summary of the working.

RohanBhirangi committed:
wrote README.md

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,29 @@
+# About in-toto
+A software supply chain is the series of steps performed when writing, testing, packaging, and distributing software. A typical software supply chain is composed of multiple steps "chained" together that transform (e.g., compilation) or verify the state (e.g., linting) of the project in order to drive it to a final product.
+
+In simple words, in-toto guarantees that the end-user (or owner) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (participants) have perfomed the specified tasks and there hasn't been any malicious changes in the files.
```
I think this could be split in a couple of shorter, simpler sentences.

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,29 @@
+We have an open-source implementation, which is still under heavy development.
+You can clone our repository and run in-toto commands (run, verify, etc.) from command line.
+
+# Contributing to in-toto
```
Thanks for adding this section! it definitely goes in, but it's weird to have it just empty with nothing in it right?

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,29 @@
+
+In simple words, in-toto guarantees that the end-user (or owner) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (participants) have perfomed the specified tasks and there hasn't been any malicious changes in the files.
+
+So what in-toto basically does is that it asks for a project layout (JSON) stating the layout's expiry, functionary keys and steps involved. The project owner is tasked with layout specification. The layout also specifies the functionaries and the tasks they are supposed to perform.
```
I'd personally drop the (JSON) part, considering that maybe we'll move on to another standard.

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,29 @@
+
+# Using in-toto
+We have an open-source implementation, which is still under heavy development.
+You can clone our repository and run in-toto commands (run, verify, etc.) from command line.
```
This reads a little bit odd, given that's the README of the OSS implementation itself :P. Maybe we could say something else here?

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,29 @@
+
+So what in-toto basically does is that it asks for a project layout (JSON) stating the layout's expiry, functionary keys and steps involved. The project owner is tasked with layout specification. The layout also specifies the functionaries and the tasks they are supposed to perform.
+After each functionary performs its task, a link metadata is generated.
+This metadata forms a part of the final product where it is used by the client to verify that there hasn't been any changes in the files between different steps.s
```
Is this incomplete? there's a spureous s at the end.

SantiagoTorres reviewed (changes requested):
Thanks for taking care of this! Overall, I think it's in a good direction, but I left some small comments that you can use to tweak this PR.

RohanBhirangi committed:
fixed README.md

RohanBhirangi committed:
fixed README.md

RohanBhirangi committed:
fixed README.md

lukpueh commented on README.md:
```diff
@@ -1,2 +1,49 @@
+This metadata is used to verify the intermediate and final products with the project layout.
+
+# in-toto demo
+You can try in-toto by running the demo application at [https://github.com/in-toto/in-toto/tree/develop/demo].
```
Please use GitHub's markdown link markup, e.g. `[demo application](https://github.com/in-toto/in-toto/tree/develop/demo)`

lukpueh commented on README.md:
```diff
@@ -1,2 +1,49 @@
+- Bob writes a python script and in-toto hashes the product and stores it to a piece of link metadata. The metadata is signed with Bob's private key. Bob sends the python script to Carl for packaging.
+- Carl packages the script into a tarball and in-toto stores the hash to the metadata.
+- The client (end-user) can now verify the final product which contains the layout, tarball, the both link metadata. The in-toto-verify command basically checks whether the layout hasn't expried and if the steps were performed according to the layout definitions.
+
```
IMHO the demo does not need to be explained here, I think the demo README already gives enough context. Just linking to the demo should be fine.

lukpueh commented on README.md:
```diff
@@ -1,2 +1,49 @@
 in-toto is a series of scripts to protect the supply chain integrity.
+
+# About in-toto
+in-toto guarantees that the end-user (or owner) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (participants) have perfomed the specified tasks and there haven't been any malicious changes in the files.
```
I'd rather not call the end-user "owner", since we are already referring to the actor who writes and signs an in-toto layout as "(project) owner". Also, instead of participants you I'd prefer "(e.g. developers)" or "(carrying out the steps of a software supply chain)".

lukpueh commented on README.md:
```diff
@@ -1,2 +1,49 @@
+# Using in-toto
+* Clone our [repository](https://github.com/in-toto/in-toto.git).
+* Run setup.py
+* Start using the in-toto run and verify scripts
```
Maybe you could call this section something like "Getting started" with the following sub sections:

1. **Installation**
similar to what you wrote in `# Contributing to in-toto`
2. **Create layout**
This will be rather complex. Maybe you can start with a simple description of the parts a layout must/should contain (steps, inspections, functionary keys) and the basic matchrule grammar?
3. **Perform software supply chain steps**
A brief description of the command line tools, `in-toto-run` and `in-toto-record`. What they do and how to use them.
4. **Release final product**
We do not strictly specify how in-toto metadata files should be shipped out together with the final product. For now this section should mention something like:
*In order to verify the final product with in-toto, the verifier must have access to the `layout`, the `*.link` files and the project owner's `public key`.*
5. **Verify final product**
A brief description of the command line tool, `in-toto-verify`.

lukpueh commented on README.md:
```diff
@@ -1,2 +1,49 @@
+*requirements file*
+pip install pycrypto cryptography
+
+Submit a Pull Request and start contributing to in-toto.
```
Maybe we can leave out this section for now. Can you please move the installation instructions up to `# Using in-toto` (`# Getting started/Insallation`)?
Also, it would be great if you could run the installation on a vanilla linux user machine (e.g. ubuntu) and find out which system packages are required to install `pycrypto` and `cryptography`.

lukpueh reviewed (dismissed):
Thanks, @RohanBhirangi for the README. It is a great start and I appreciate that you are working on this. See my improvement suggestions below.

lukpueh commented on README.md:
```diff
@@ -1,2 +1,54 @@
+
+pip install pycrypto cryptography
+
+Submit a Pull Request and start contributing to in-toto.
```
Maybe we can leave out this section for now. Can you please move the installation instructions up to `# Using in-toto` (`# Getting started/Installation`)?
Also, it would be great if you could run the installation on a vanilla linux user machine (e.g. ubuntu) and find out which system packages are required to install `pycrypto` and `cryptography`.

RohanBhirangi committed:
Fixed README.md

RohanBhirangi committed:
Fixed README.md

RohanBhirangi committed:
Fixed README.md

RohanBhirangi committed:
Fixed README.md

RohanBhirangi committed:
Fixed README.md

SantiagoTorres commented:
As a general comment: it's really helpful if you use more descriptive commit messages. You can look at the log to see how we are doing them. I know it's a little counter-intuitive or confusing at first, but you'll get better at it with time 👍

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,83 @@
 in-toto is a series of scripts to protect the supply chain integrity.
+
+# About in-toto
+in-toto guarantees that the end-user (or client) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (eg. developers) have performed the specified tasks and there haven't been any malicious changes in the files.
```
I think we could avoid the use of layout without explaining what it is. Overall this feels ok though.

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,83 @@
+# About in-toto
+in-toto guarantees that the end-user (or client) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (eg. developers) have performed the specified tasks and there haven't been any malicious changes in the files.
+
+in-toto requires a project layout that specifies the functionaries and the tasks they are supposed to perform.
```
I would rephrase as:
> in-toto let's project owners to define which steps should be performed and who can perform them to derive a final product.

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,83 @@
+in-toto guarantees that the end-user (or client) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (eg. developers) have performed the specified tasks and there haven't been any malicious changes in the files.
+
+in-toto requires a project layout that specifies the functionaries and the tasks they are supposed to perform.
+After each functionary performs its task, a link metadata is generated.
```
`a piece of link metadata is generated` or `a link metadata file is generated` or simply `link metadata is generated`

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,83 @@
+
+in-toto requires a project layout that specifies the functionaries and the tasks they are supposed to perform.
+After each functionary performs its task, a link metadata is generated.
+This metadata is used to verify the intermediate and final products with the project layout.
```
This feels a little bit ambiguous. What about:

> This metadata file is used to verify that the materials used and the products produced within this task weren't tampered with. The link metadata also contains a signature to ensure that the right functionary performed this task.

Or something along those lines

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,83 @@
+This metadata is used to verify the intermediate and final products with the project layout.
+
+# in-toto demo
+You can try in-toto by running the demo application at [demo application](https://github.com/in-toto/in-toto/tree/develop/demo).
```
This is ok, but maybe we can delete the repeated link:

> You can try in-toto by running the [demo application](https://github.com/in-toto/in-toto/tree/develop/demo).

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,83 @@
+*Make sure you have git, python and pip installed on your system*
+*and get in-toto*
+
+git clone -b develop --recursive https://github.com/in-toto/in-toto.git
```
I would:

1.- put this in a code block (you can use  \`\`\` to start and end a code block
2.- Now that the develop branch is the default branch, we don't need the `-b develop` switch. We also need to correct this in the demo text.

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,83 @@
+  5. "signature": which denotes the private key of the owner
+  6. "steps": which describes the steps involved and the functionaries who are authorized to perform them
+
+You can use the create_layout.py script or write your own script to specify a layout for your project.
```
I like this. Maybe we could link to the create_layout.py script in here also.

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,83 @@
+      * the recorded materials and products align with the matchrules and
+      * the inspection untar finds what it expects.
+
+# Specifications
```
I would make this singular.

SantiagoTorres reviewed (changes requested):
Some more small notes. It's getting there!

Commit:
Updates README.md (various style and content)

- Adds some verbatim github markdown markup and bullet points
- Moves demo link to the end of document
- Adds Install Dependencies (tested on a fresh ubuntu desktop)
- Rewords some things
- Adds command help messages

lukpueh commented:
Thanks @RohanBhirangi, for your solid groundwork! I permitted myself to take it a little further before we merge.

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,106 @@
-in-toto is a series of scripts to protect the supply chain integrity.
+Protecting the software supply chain integrity
+
+in-toto guarantees that the end-user (or client) is able to verify that the entire development life cycle has been conducted as per the specified layout and that each of the functionaries (eg. developers) have performed the specified tasks and there haven't been any malicious changes in the files.
```
I don't know if we should assume that the term "layout" is self-explanatory. What do you guys think?

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,106 @@
+
+### Install Dependencies
+ - [Python](www.python.org) in version 2.7 - crypto libraries require header files
+ - [OpenSSL](https://www.openssl.org/) - crypto libraries require header files
```
Do we require openssl? who uses openssl as a backend?

SantiagoTorres replied:
should we instead mention openssl-dev and stuff like that?

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,106 @@
+ - [OpenSSL](https://www.openssl.org/) - crypto libraries require header files
+ - [git](https://git-scm.com/) - version control system
+ - [pip](https://pip.pypa.io) - package installer tool
+ - [virtualenvs](http://docs.python-guide.org/en/latest/dev/virtualenvs/) - optional but strongly recommended
```
didn't we have a guide on venv + venvwrapper?

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,106 @@
+cd in-toto
+
+# Install with pip in "develop mode"
+pip install -e .
```
Technicallity, but I think it is called "editable" mode. Maybe we want to expand to "--editable" given that we use "--recursive" above.

SantiagoTorres commented on README.md:
```diff
@@ -1,2 +1,106 @@
+            [--materials <filepath>[ <filepath> ...]]
+            [--products <filepath>[ <filepath> ...]]
+            [--record-byproducts]
+            [--verbose] -- <cmd> [args]
```
Maybe we don't need all the flags here? If so, I think we are missing a couple.

SantiagoTorres reviewed (changes requested):
Looks good!
